### PR TITLE
Update method now runs a single UPDATE query on mongoDB

### PR DESCRIPTION
**Pull request by amobarak**
Fixed DSND-376 database persistence issue, improves database update performance (1 call instead of 2 calls before).
Now also updates Etag

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java [hunk 1]
```diff
@@ -1,7 +1,13 @@
 package uk.gov.companieshouse.company.profile.service;
 
+import com.mongodb.client.result.UpdateResult;
 import java.util.NoSuchElementException;
 import java.util.Optional;
+import org.springframework.beans.factory.annotation.Autowired;
```
Comment (iichr) [inline src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java:6]: Unused import, but for some reason checkstyle is not flagging it for me either - IntelliJ is

File: src/test/java/uk/gov/companieshouse/company/profile/service/CompanyProfileServiceTest.java [hunk 2]
```diff
@@ -7,20 +7,19 @@
 import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.ArgumentMatchers.argThat;
 import static org.mockito.ArgumentMatchers.eq;
-import static org.mockito.Mockito.when;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.*;
```
Comment (iichr) [inline src/test/java/uk/gov/companieshouse/company/profile/service/CompanyProfileServiceTest.java:10]: no star imports

**Review by iichr — DISMISSED** [review]

File: src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java [hunk 3]
```diff
@@ -42,19 +54,21 @@ public CompanyProfileService(Logger logger, CompanyProfileRepository companyProf
...
     public void updateInsolvencyLink(final CompanyProfile companyProfileRequest)
             throws NoSuchElementException {
-        Optional<CompanyProfileDocument> companyProfileOptional = companyProfileRepository
-                .findCompanyProfileDaoByCompanyProfile_Data_CompanyNumber(
-                        companyProfileRequest.getData().getCompanyNumber());
+        String companyNumber = companyProfileRequest.getData().getCompanyNumber();
+        Query updateCriteria = new Query(Criteria.where("data.company_number").is(companyNumber));
+        Update updateQuery = new Update();
+        updateQuery.set("data.links.insolvency",
+                companyProfileRequest.getData().getLinks().getInsolvency());
```
Comment (zahidm3) [inline src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java:61]: I think we also need to update etag, updatedBy along with insolvency links element? I would probably prefer to fetching whole object once and persist object with all updated attributes maybe

File: src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java [hunk 4]
```diff
@@ -42,19 +54,21 @@ public CompanyProfileService(Logger logger, CompanyProfileRepository companyProf
...
+                updateQuery,
+                "company_profile");
 
-        if (companyProfileOptional.isEmpty()) {
-            throw new NoSuchElementException("Database entry not found");
+        if (updateResult.getModifiedCount() == 1) {
+            logger.trace(String.format("DSND-376: Insolvency links updated for company number: %s",
+                    companyNumber));
+        }
+        if (updateResult.getMatchedCount() == 0) {
```
Comment (zahidm3) [inline src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java:70]: I dont think this approach will work once we start updating multiple attributes only if the links need to be updated

**Review by zahidm3 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java (hunk 3 above)
Comment (amobarak) [reply src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java:61]: I wasn't aware of updating etag/updatedBy, would it be better to update the update query to include these two fields, to keep it as a single DB call

File: src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java (hunk 4 above)
Comment (amobarak) [reply src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java:70]: I thought we are updating a single document as we are fetching one by its companyNumber (Where companyNumber is its ID too)

File: src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java [hunk 5]
```diff
@@ -61,25 +67,32 @@ public CompanyProfileService(Logger logger, CompanyProfileRepository companyProf
...
+        if (updateResult.getModifiedCount() == 1) {
+            logger.trace(String.format("DSND-376: Insolvency links updated for company number: %s",
+                    companyNumber));
+            try {
+                insolvencyApiService.invokeChsKafkaApi(companyNumber);
+                logger.info(String.format("DSND-377: ChsKafka api invoked successfully for company "
+                        + "number %s", companyNumber));
+            } catch (Exception exception) {
+                logger.info(String.format("Error invoking ChsKafka API for company number %s",
+                        companyNumber));
```
Comment (iichr) [inline src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java:90]: guessing it's needed even if the invoke itself has a try catch inside?

**Review by iichr — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java (hunk 5 above)
Comment (amobarak) [reply src/main/java/uk/gov/companieshouse/company/profile/service/CompanyProfileService.java:90]: Yes, a log to trace if kafka message failed

**Review by zahidm3 — APPROVED** [review]

**Pull request merged**